tokio-trace/tokio-trace-core/src/metadata.rs

```diff
@@ -122,6 +122,25 @@ pub struct Metadata<'a> {
     /// `Metadata::new` constructor instead!
     #[doc(hidden)]
     pub fields: field::FieldSet,
+
+    /// The kind of the callsite.
+    ///
+    /// **Warning**: The fields on this type are currently `pub` because it must
+    /// be able to be constructed statically by macros. However, when `const
+    /// fn`s are available on stable Rust, this will no longer be necessary.
+    /// Thus, these fields are *not* considered stable public API, and they may
+    /// change warning. Do not rely on any fields on `Metadata`. When
+    /// constructing new `Metadata`, use the `metadata!` macro or the
+    /// `Metadata::new` constructor instead!
+    #[doc(hidden)]
+    pub callsite_kind: CallsiteKind,
+}
+
+/// Indicate whether the callsite is a span or event.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub enum CallsiteKind {
```

Comment:
I'd prefer not to have enums exposed publicly, since this presents a forwards-compatibility hazard (enum matching is exhaustive, and adding new variants will break any match statements). I'd like to reserve the possibility of potentially adding more kinds of callsites without causing a breaking change. 

As I see it, we essentially have two options for this. If we want to continue with the current API, where `Metadata` is constructed by a single `new` function with a `CallsiteKind` provided as an argument, we would want to use a similar strategy to how the `Level` type is currently implemented, as a public struct wrapping a private enum. We would expose `CallsiteKind::event()` and `CallsiteKind::span()` functions to construct the variants (or associated consts, like `Level`), and `CallsiteKind::is_event`/`is_span` functions. 

Alternatively, we could just split `Metadata::new` into separate `Metadata::new_span` and `Metadata::new_event` constructors, which could set the kind field. Then, it would be safe to just have a `CallsiteKind` enum. We would probably want it to be `#[doc(hidden)] pub` in that case, since it would still need to be used by the `metadata!` macro, but we could treat it as though it was private.